Implement the new permissions policy

rijkvanzanten opened the pull request:
## Todos


- [ ] Add permissions merging when you're accessing from a share
- [ ] Bring back check-ip middlewares?
- [ ] Fix `down` migration
- [x] Add changesets
- [x] Add permissions processing for $CURRENT_USER etc flags
  - [x] Introduce $CURRENT_POLICIES and $CURRENT_ROLES for permissions
  - [x] Decide if $CURRENT_POLICIES and $CURRENT_ROLES should be available in presets as well (Decided: they won't for now)
- [x] Use applyCases in Meta Service for permissions aware counts
- [x] Purge request level cache on permission related changes 
- [x] Reduce accountability used in the `withCache` to the known keys (Handled in https://github.com/directus/directus/pull/22773/commits/2e156f549a009435141d64e268ef58e1bc38ebc9)
- [x] Fix handling of o2m fields that have partial permissions set (currently if some access to a o2m related field is given, it is returned for all items, even if the user should not see the field for some items) (Fixed in https://github.com/directus/directus/pull/22773/commits/47a9edc9a4247cfdbf3992f551d964f960e16ca8)
- [ ] Fix explicit requests of o2m fields where the user does not have access to the related collection. (Right now this is silently dropped when generating the AST in order to avoid errors when requesting `*`).
- [x] Account for user IP in the global request cache, but only if the IP matches any ip_access set in any policy (Fixed in https://github.com/directus/directus/pull/22773/commits/3ab3360c7c8441a917f71c08fcdf3ec2e372a8a4)
- [x] Make sure that the globalAccess for a public user always has `app_access` and `admin_access` `false`  (This is actually already accounted for, since the public user just receives the default accountability)
- [x] Remove fixed public policy from the app
- [x] Add icon field to policies
- [x] Prevent infinite role recursion 
- [x] Add new collections to GraphQL (See #22795)
- [x] Add new collections to websockets (See #22795)
- [x] Add new endpoints + types to SDK (See #22795)
- [x] Merge sdk/gql updates (See #22795)
- [x] Merge doc updates (See #22729)
- [x] Rename migration to the latest date in order to ensure down migration order

rijkvanzanten committed:
WIP start on migrations

rijkvanzanten committed:
Add migration

rijkvanzanten committed:
Don't insert if there's no rows

rijkvanzanten committed:
Use service to read/write permissions

rijkvanzanten committed:
Use payload service rather than itemsservice

rijkvanzanten committed:
Start on downgrade command

rijkvanzanten committed:
Update system data structure

rijkvanzanten committed:
Update migrations to keep structure flat

rijkvanzanten committed:
Remove icon from policies

rijkvanzanten committed:
Drop policies table on downgrade

rijkvanzanten committed:
Rearchitect migrations to structure v3

rijkvanzanten committed:
Add down migration

rijkvanzanten committed:
Update system fields

rijkvanzanten committed:
Add policy to fields import

rijkvanzanten committed:
Fix public role attachment

rijkvanzanten committed:
Update packages/system-data/src/fields/index.ts

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>

rijkvanzanten committed:
Update packages/system-data/src/fields/policies.yaml

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>

rijkvanzanten committed:
Add nested roles

rijkvanzanten committed:
Remove unused step

rijkvanzanten committed:
Use two o2ms instead of m2a for attachments

rijkvanzanten committed:
Update system data

rijkvanzanten committed:
Merge branch 'main' into auditus

rijkvanzanten committed:
Merge branch 'main' into auditus

rijkvanzanten committed:
Merge branch 'main' into auditus

rijkvanzanten committed:
Merge branch 'main' into auditus

rijkvanzanten committed:
Merge branch 'main' into auditus

rijkvanzanten committed:
Merge branch 'main' into auditus

rijkvanzanten committed:
Merge branch 'main' into auditus

rijkvanzanten committed:
Merge branch 'main' into auditus

rijkvanzanten committed:
Merge branch 'main' into auditus

rijkvanzanten committed:
Merge branch 'main' into auditus

rijkvanzanten committed:
Merge branch 'main' into auditus

hanneskuettner committed:
Merge branch 'refs/heads/main' into auditus

hanneskuettner committed:
Merge branch 'main' into auditus

rijkvanzanten committed:
Merge branch 'main' into auditus

hanneskuettner committed:
Merge branch 'main' into auditus

rijkvanzanten committed:
Implement permission policies in the API (#22384)

* Update system data structure

* Update migrations to keep structure flat

* Remove icon from policies

* Drop policies table on downgrade

* Rearchitect migrations to structure v3

* Add down migration

* Update system fields

* Add policy to fields import

* Fix public role attachment

* Update packages/system-data/src/fields/index.ts

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>

* Update packages/system-data/src/fields/policies.yaml

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>

* Add nested roles

* Remove unused step

* Use two o2ms instead of m2a for attachments

* Update system data

* [WIP] Start reorging permissions handling

* Setup field extraction

* Remove watch from vitest

* Finish fieldMap creation logic

* Add tests for utils

* Improve tests

* Improve coverage

* Split test and test:watch

* Continue on this fun

* [WIP] Setup processing

* Sort roles

* Restructure to util files for org

* Add missing util tests

* More tests

* Add cases/whencase to ast

* Start on injection logic

* Add tests for inject cases

* Add tests for process

* Add todo

* Organize run-ast

* Add clear method to kv

* Remove reliance on acc.perm

* Restructure permissions setup

* Drop perm from acc, add roles/policies

* Remove get-permissions in middleware

* Remove/comment use of acc.perm

* Add default roles/permissions

* Use knex

So we don't have to initialize the schema before we want to use the accountability system

* Use new fetching logic in get accountability

* Add new fetch global access utils

* Gotta redo based on new setup

* Replaced with new util

* Remove dropping of perm in acc

It's no longer there by default, so no need to remove here

* Temporarily comment out the enforce tfa check

* Update usage of fetch tree to use knex

* Don't store policies on accountability

* Feed in roles thru acc

* Bit of whitespace

* Rename role->policy

* Wreck some more stuff

Jk, this is splitting up the large get-ast-from-query function into smaller individual functions to make it easier to update the wildcard conversion to use permissions

* Add ability to lookup all allowed fields in col+ac

* Add note so I don't forget stuff which i will

* Handle null acc

* Introduce parseAst to itemsservice

* That cleans things up

* Replace checkAccess with validateAccess

* Remove checkaccess from service

* cleanup imports

* Whoops one more

* Leave crumbs for next time

* Implement most of the fn

* Fix various tests

* Start on test for fetch roles tree

* Add tests for fetch roles tree

* Fix process tests

* All. of. the. tests.

* Update uses of validateAccess

* Fix name in runAst

* Fix use of accountability in gql sub

* Deprecate authorization service

* Remove getPermissions use

* Drop old getpermissions

* Pass services

* Replace admin/app uses with fetch global

* Update fetch user count to pull from policies

* Remove broken admin existence checks

* Update min accountability

* Remove unused import

* Drop permissions override from controller

* Refactor reliance on acc.perm

* Replace usage of permissions in fields

* Replace usage of permissions in import/export

* Drop permissions use from relations

* Drop no longer used method

* Remove unused import

* fix type usage of pk in validate

* Fix default acc in user

* Replace use of permissions in utils

* Update reduceSchema in specs/gql

* Remove old share merging

* Remove empty file

* Remove outdated comment

* Use ctx objects for large param fns

* Add with-cache memoize util

* Add cache to fetchpermissions

* Update caching use in fetchRolesTree

* Add caching to fetchAllowedFieldMap

* Add more cache

* Refactor call signatures

* Move call signature updates

* Handle presets

* Update process call sig

* Prevent infinite recursion in roles tree lookup

* Use create util for acc

* Remove old checkIp

* Fix where equality operator

* Break EVERYTHING!

Jk just cleaning up the structure some more, and removing the dep injection in favor of mocking

* Fix build

* Add missing module tests

* Don't crash on missing parent

* Fix role lookup

* add missing type annotation

* use logical-OR assignment and avoid a memory allocation

* Attach admin policy in default admin creation

* Fix admin check

* Add todo for later

* rm code duplication

* fix test

it was missing the new `roles`

* add types and fix type error

policies dont [yet] have an icon

* move spread order to avoid potential future mishaps

new default keys would override the manually set keys, potentially leading to unintended behavior

* reduce allocations, add escape hatch to loop and type db-row

* Implement case/when

* Clean up comments

* Optimize perm fetching in allowed f

* Move apply case when to util fn

* Optimize fetch-allowed-fields

* Add fetch inconsistent util

* Allow nulls

* Remove obsolete getCacheKey

* Remove unused import

* Update getAccountabilityForRole test

* Update fetchGlobalAccess test with one more test case + fix other test case

* Type cleanup

* Fix "admin access means automatic app access" in fetchGlobalAccessForQuery

* Clean up and expand fetch-inconsistent-field-map.test.ts

* Test uncached functions

* Test uncached

* Remove cases usage in parse-current-level

* Only consider non-null rules in inject cases

* Fix parseCurrentLevel call

* Move service imports into functions to avoid circular imports

* Ensure that we test that an error is thrown in processAst test

* Add failing test case for flattenFilter

* Ensure uniqueness in extractPathsFromQuery

* Early exit in validatePath

* Add additional test case for process payload test

* Update validateCollectionAccess test

* Clean up validate-item-access.test.ts

* Remove redundant initializer

* Use createDefaultAccountability

* Fix fetch-user-count.test.ts

* Cleanup unused default initializer

* Add empty cases to subfilter in _relationCount

* Drop AccessService and PermissionsService usage from services

* Found some more PermissionsServices

* Fix a few more tests

* Add nested role relation

* Fix query invocation in aggregate and group queries

* Fix role property name in auth/refresh

* Add some missing relations for permissions, access and roles

* Add m2o relation from permissions to policy

* Add m2o relation access to role, user, policy

* Allow fetchPermissions to fetch all permissions and not just those limited by an action

* Add parent to Role type

* Make sure that admin users see all fields

* Add access and policies controller, add util methods to policies and access service

* Change name and description of public policy, update description of admin policy and add on delete trigger.

* Make sure access row uuids are auto generated

* optimize kvredis clear function and add a unit test

to be fair: unit test is also testing implementation details but thats a problem there in general and for future us

* Add minimal app permission and dynamic variable injection to the permission fetching

* Fix m2o collection name in extractFieldsFromChildren

* Make sure to clone permission before injecting dynamic variables

* Actually do the cloning in with withAppMinimalPermissions since people might missbehave with the permissions obtained from PermissionsService.readByQuery so it better to go the source of the problem

* Use knex transaction in createOne -> processPayload - otherwise deadlock

* Make sure to respect '*' field in allowed fields

* Fix extractFieldsFromChildren for o2m as well - classic

* Fix allowed field check in `FieldsService.readAll` to account for multiple permissions for collection+action

* Skip case/when if `allowedFields` includes '*'

* Restructure the way the current users permissions are returned

* add ability to clear all keys from memory cache

* add test for clear method

* add await to clear function

* Clear permissions caches on changes to policy attachments (directus_access) and policy updates (directus_policies) and permissions updates (directus_permissions)

* Make the public role a real role rather than a virtual one

* Inject the public role, we're it previously was `null`

* Revert adding a fix public role

* remove unused variable

* Ensure that a user without a role can still use the /me util endpoints

* Make sure that the /me endpoints always return minimal information, similar to /users/me

* Some fixes after merging main

* Update api/src/permissions/utils/with-cache.ts

Co-authored-by: Hannes Küttner <4376726+hanneskuettner@users.noreply.github.com>

* Avoid broken role query for now

* Skip related collection `parseFields` if user has no permissions

* Ensure same call order as in `convertWildcards`

* Create default admin policy and connect it in cli init command

* Remove obsolete middleware mock in app.test.ts

* Add validation against non-existent fields and collections to `validatePath`

* Split up permission and path existence validation and validate path existence for admin users as well

* Make applySearch not async

* Fix relation extraction and permissions for `$FOLLOW` fields

* Fix case when for related collections and query wrapping

* Rework user integrity checks for Auditus (#22737)

* Changes to user counting and integrity checks

* Ensure that user validation happens in both create one and create many

* Rename `checkType` to `flags`

* Update api/src/permissions/modules/validate-remaining-admin/validate-remaining-admin-count.ts

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>

* Update to enum usage

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>

* A few more changes to enum instead of number

* One more enum type update

* Make sure to correctly override the callback when combining options

* Clean up option type

* Update api/src/services/users.ts

Co-authored-by: ian <licitdev@gmail.com>

* Only take validation shortcut for users

We can only be sure that the deletion of users does not increase any other access types count, so in all other cases we need to verify that for example the App or API users have not increased over the limit

* Make both app and admin users count against app access limit

* Update api/src/permissions/modules/validate-remaining-admin/validate-remaining-admin-count.ts

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

* One post-merge fix, two small fixes

* Simplify flag updating and callback calling

* Changing app access in a policy only requires user limit checking, not full check

* Only the status of a created user should matter to determine if a check is neccessary

* Add count alias to count query

---------

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>
Co-authored-by: ian <licitdev@gmail.com>
Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>
Co-authored-by: Rijk van Zanten <rijkvanzanten@me.com>

* Add roles and permissions to the app (#22654)

* Initial app changes

* Fix getRelationsForField

* Add changeset

* Remove app-permissions from role settings

* Make sure access row uuids are auto generated

* Move a few things around, set up policies m2m properly

* Show roles as tree in sidebar
Change avatar field query for user

* Show user and role count in policy table

* Default to not adding app access for a policy, makes composability less annoying

* Correctly fall back to 0 for counts

* Change the structure of current user permissions

* Start bringing back the public role

* Make the public role a real role rather than a virtual one

* Revert public role changes

* Extend list-m2m to allow for very custom junction matching and a primary key of `null`

* Remove unused

* Fix public role policy update payload

* Fix app access for users without role (which is a thing now apparently)

* Make sure that the /me endpoints always return minimal information, similar to /users/me

* Tweak nav icons

* Pull policy id from constants

* Update permissions interface design to match

New design language in figma

* Some minor adjustments

- Make chip hover border more consistent
- Add "Remove" button to remove a full row of permissions, as in the UI mockup
- Fix table layout

* Clean up a few more things

* Fix `setFullAccess`

* Align collection view icons with navigation

* Don't query 'admin_access' for role

* Fix relation extraction and permissions for `$FOLLOW` fields

* Don't show `0 Items` for child rows, but `--` instead

* Make policy detail work in nested policy creating use case

* Remove unused v-icon override

* Move system collections to separate visual table

* Navigate before refresh

Prevents a flash of the previous value to be visible in the table

* Move composable to separate file

---------

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>
Co-authored-by: Rijk van Zanten <rijkvanzanten@me.com>

* Optimize types

* Clone query deep

* Optimize type order

* Throw error on invalid role id

* Rename run.js -> run-ast.js

* Re-add filesizes to telemetry report collection that got lost in the merge

* Make `systemCollections` reactive

* Use one column per action to avoid unwanted shifting if some actions are not allowed at all

* Render system and custom together

* Add divider between regular and system permissions if both have elements

* Add AccessService and PoliciesService to `getService`

* Move policy global flags fetching to util

* Move collection access fetching into util

* Remove permissions for `directus_access`, `directus_permissions` and `directus_policies` from schema permissions

* use formatted-value display for name & description in roles & policies

* Rename `process.ts` to `process-ast.ts`

* Fix process-ast import after renaming

* Perform user integrity check on item deletion

* Fix first admin creation on bootstrap

* Revert "Fix first admin creation on bootstrap"

This reverts commit bf480d023c508a93e6d7304ae8f52c70af47da7e.

Will be fixed by adjusting the check in access service

* Don't perform admin integrity check if a new access row is created. Only check user limits

* Don't set an alias to the raw column value if it is wrapped in a case/when

* Correctly handle aliases when in field map and case injection

---------

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>
Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>
Co-authored-by: Hannes Küttner <kuettner.hannes@gmail.com>
Co-authored-by: Hannes Küttner <4376726+hanneskuettner@users.noreply.github.com>
Co-authored-by: ian <licitdev@gmail.com>

hanneskuettner committed:
Remove changeset of already merged PR (#22653)

hanneskuettner committed:
Fix multi cache subscribe call to preserve context

Co-authored-by: Brainslug <br41nslug@users.noreply.github.com>

br41nslug commented on app/src/modules/settings/routes/roles/add-new.vue:
```diff
@@ -65,16 +50,6 @@ function useSave() {
 					<div class="field full">
 						<v-input v-model="roleName" autofocus :placeholder="t('role_name') + '...'" @keyup.enter="save" />
```
The UI does not correctly limit the user input to the maximum length that the backend will accept (100 chars) leading to a cryptic error if attempting  a longer name.
```suggestion
						<v-input v-model="roleName" autofocus :placeholder="t('role_name') + '...'" @keyup.enter="save" maxLength="100" />
```
![firefox_UO7HVFi9A2](https://github.com/directus/directus/assets/9389634/be1210ec-cb0b-48bf-b37e-8cc4bbb963da)

br41nslug commented on app/src/modules/settings/routes/policies/add-new.vue:
```diff
@@ -0,0 +1,61 @@
+			<v-card-text>
+				<div class="form-grid">
+					<div class="field full">
+						<v-input v-model="name" autofocus :placeholder="t('policy_name') + '...'" @keyup.enter="save" />
```
This input is not limited same as the role name input should have a maximum length in sync with what the backend will accept (100 chars) 
```suggestion
						<v-input v-model="name" autofocus :placeholder="t('policy_name') + '...'" @keyup.enter="save"  maxLength="100 />
```

hanneskuettner committed:
Add max length to name for policy and role names

Co-authored-by: Brainslug <br41nslug@users.noreply.github.com>

br41nslug commented on app/src/modules/settings/routes/policies/collection.vue:
```diff
@@ -0,0 +1,262 @@
+					<v-icon class="icon" :name="item.icon" :class="{ public: item.public }" />
+				</template>
+
+				<template #[`item.name`]="{ item }">
+					<v-text-overflow :text="item.name" class="name" :class="{ public: item.public }" :highlight="search" />
+				</template>
+
+				<template #[`item.userCount`]="{ item }">
+					<value-null v-if="item.public" />
+				</template>
+
+				<template #[`item.roleCount`]="{ item }">
+					<value-null v-if="item.public" />
+				</template>
+
+				<template #[`item.description`]="{ item }">
+					<v-text-overflow :text="item.description" class="description" :highlight="search" />
+				</template>
```
When searching through a large list of policies im getting the following error where it tries to do things on a `string` that is in fact a `null`.
![firefox_4UEBMacM3u](https://github.com/directus/directus/assets/9389634/ea3747d1-87e5-4dda-a67a-d90a27ea4cf7)

```suggestion
				<template #[`item.name`]="{ item }">
					<v-text-overflow v-if="item.name" :text="item.name" class="name" :class="{ public: item.public }" :highlight="search" />
				</template>

				<template #[`item.userCount`]="{ item }">
					<value-null v-if="item.public" />
				</template>

				<template #[`item.roleCount`]="{ item }">
					<value-null v-if="item.public" />
				</template>

				<template #[`item.description`]="{ item }">
					<v-text-overflow v-if="item.description" :text="item.description" class="description" :highlight="search" />
				</template>
```

hanneskuettner committed:
Update app/src/modules/settings/routes/policies/collection.vue

Co-authored-by: Brainslug <br41nslug@users.noreply.github.com>

br41nslug commented on app/src/modules/settings/routes/roles/item.vue:
When opening the role detail page a fetch call goes out attempting to count the roles.
On postgres this critically fails when there are no nested roles (or role.children):

![image](https://github.com/directus/directus/assets/9389634/fd3df6ab-4ba0-452d-b4a3-407d60b0de2f)

Caused by the empty `_nin` filter generated in the app
http://localhost:8080/roles?limit=25&fields[]=icon&fields[]=name&fields[]=description&fields[]=id&sort[]=name&page=1&filter[_and][0][id][_neq]=08a4e97b-ed1e-41b7-9926-606de95ca7c0&filter[_and][1][id][_nin][0]=

connorwinston replied:
Also had this happen. Prevents you from being to select a parent role as the table to select the role never loads.

hanneskuettner replied:
I'll just kick that check for now and we can re-iterate on it. I was trying to prevent a first level of infinite role recursion app side, bit this probably actually needs a api side check.

hanneskuettner replied:
Done. Should work now (don't forget to rebuild `@directus/system-data`)

br41nslug replied:
Currently the original Database error has been fixed 🎉 However when opening a role page (no policies, users or parent) i am consistently seeing this error:
![image](https://github.com/directus/directus/assets/9389634/ad68b469-22bb-4e5a-b89c-cd1c1448914a)

rijkvanzanten committed:
Organize settings sidebar for clarity

hanneskuettner committed:
Remove query limit override from access and policies controller

THe query limit was added previously, where the idea was to fetch all policies in the app, but since that is not required anymore we can remove the override again, which in turn fixed pagination for policies. Woop

hanneskuettner committed:
Rework MetaService

* Fix filtered counting (previously it did not account for left join caused by permissions filter)
* Collect permissions for current collection and dedupe access to align filter with the one used in the actual query
* Use applyFilter and an _or filter to retrieve permitted items

hanneskuettner committed:
Prevent certain skips in _or filters

We can only skip empty filters in `_or` if either they are not equal to the value of `cases` or if _or has exactly only exactly one empty filter

This is needed to prevent dropping joins that are required for the case/when construction. For example when having the permissions `_or: [ {}, {related_item: { id: 1} }]` all joins need to be retained

hanneskuettner committed:
Revert unintentional with-cache commit

hanneskuettner committed:
Remove check for id in children which fails on some DBs if no children are set

hanneskuettner committed:
Show users and roles in policy item view

hanneskuettner committed:
Update directus_access policy/roles + policy/users `one_deselect_action`

This ensures that the access rows are cleaned up when removing users or roles from the policy side of the relation

hanneskuettner committed:
Merge policy loaded from API with current edits

hanneskuettner committed:
Make `app_access` default to false

hanneskuettner committed:
Split field map into read specific and other fields

This change is necessary since process-ast is used to verify item access for actions other than `read`.
But, if a user does not have action permissions for fields used in the query filter or sort field the validation for `xByQuery` would have failed until now.

* The fields are verified separately checked against read and action specific permission.
* Updated all the tests accordingly

hanneskuettner committed:
Fix `hasCaseWhen` check in `getDBQuery`

Previously it was checking if `cases.length > 0` which was always true, since we always pass in at least one case (`{}`), now it checks if there are actually field nodes with a whenCase property

paescuj reviewed (commented):
When trying to remove permissions on system collections in a policy with "App Access" enabled, it is currently confusing, as some collections can be removed completely and some remain (because of minimal access).
If the interface stays like this (we've discussed some other approaches with minimal/recommended access), we should add some kind of indicator on the delete button on such collections, e.g. disable it.

https://github.com/directus/directus/assets/5363448/8eb4628c-143e-4727-8f63-51f288cd3591

hanneskuettner committed:
Don't expose o2m fields that the user might not have access to for some items

This approach uses a flag that is introduced into the parent item db query, that uses the case/when construct to determine if a user has access to the o2m field on the specific item.
The flag is 1 if the user has access and null otherwise.
It is set in the resulting query object for all o2m fields that have a whenCase (the ones with partial access) and used when merging the nested query items into the parent items.

paescuj committed:
Accept O2MNode as fieldNode

hanneskuettner committed:
Filter policies in fetchGlobalAccess by ip_access if applicable and use `withCache` util

* Filter the policies influencing the global access by ip_access filter if an ip is available
* Use `withCache` util for top level function and the two lower level functions

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>

hanneskuettner committed:
Fix filter in roles. Again. Almost like I didn't properly test it. huh

paescuj reviewed (commented):
Due to the schema changes in `directus_roles` and `directus_permissions` collections, existing permissions may no longer apply / be valid.

For example, such a negative check would now always pass since the `ip_access` field has been moved to `directus_policies`:

<img width="800" src="https://github.com/directus/directus/assets/5363448/166e5aaa-1873-4754-944d-f588ee573435">

Or field permission on moved fields, which, after migration, will still show the "Field Permissions" tab as active, while no field selected (and cannot be deselected via Data Studio, only via API):

<img width="800" src="https://github.com/directus/directus/assets/5363448/e2120e48-9487-4b72-9fc8-846cbab7f6da">
<img width="800" src="https://github.com/directus/directus/assets/5363448/38675d83-dede-4e0e-8d27-c0eaf5052cd9">

I'm afraid it would be too complicated to auto-migrate these permissions, but we have to point it out via list of breaking changes.

DanielBiegler reviewed (commented):
Currently we add `"sort":["id"]` to the AST in `ItemsService.readByQuery` which results in not being able to read items for which you dont have read access on `id` for

```
api dev: AST {"type":"root","name":"directus_files","query":{"sort":["id"]},"children":[{"type":"field","name":"title","fieldKey":"title","whenCase":[]}],"cases":[]}
api dev: [14:04:09.166] DEBUG: You don't have permission to access field "id" in collection "directus_files" or it does not exist. Queried in root.
api dev:     err: {
api dev:       "type": "",
api dev:       "message": "You don't have permission to access field \"id\" in collection \"directus_files\" or it does not exist. Queried in root.",
api dev:       "stack":
api dev:           DirectusError: You don't have permission to access field "id" in collection "directus_files" or it does not exist. Queried in root.
api dev:               at createFieldsForbiddenError (/api/src/permissions/modules/process-ast/utils/validate-path/create-error.ts:16:9)
api dev:               at validatePathPermissions (/api/src/permissions/modules/process-ast/utils/validate-path/validate-path-permissions.ts:41:9)
api dev:               at processAst (/api/src/permissions/modules/process-ast/process-ast.ts:61:3)
api dev:               at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
api dev:               at FilesService.readByQuery (/api/src/services/items.ts:500:9)
api dev:               at <anonymous> (/api/src/controllers/files.ts:241:12)
api dev:       "name": "DirectusError",
// .. truncated ..
api dev:       "code": "FORBIDDEN",
api dev:       "status": 403
api dev:     }
api dev: [14:04:09] GET /files?fields=title 403 15ms
```

Reproduction steps:

1. Update public role policy with field-level read-access to only file `title`, nothing else
2. Query API `http://0.0.0.0:8080/files` or `http://0.0.0.0:8080/files?fields=title`
3. You get the error

hanneskuettner committed:
Add cache key stability by only picking the props that are relevant from accountability and enforcing an order in the provided options object

licitdev reviewed (commented):
Any changes within the policy item page result in the refetching of counts and data for the `list-m2m` interfaces for users and roles.

https://github.com/directus/directus/assets/26413686/fab76f0f-951b-40ce-a3b5-2db8bb5643c8

licitdev commented on api/src/permissions/modules/process-ast/utils/validate-path/validate-path-existence.ts:
```diff
@@ -0,0 +1,20 @@
+
+	const requestedFields = Array.from(fields);
+
+	const nonExistentFields = requestedFields.filter(
+		(field) => !field.startsWith('$FOLLOW') && collectionInfo.fields[field] === undefined,
+	);
```
Fields such as `preferences_divider` do not exist in the DB, resulting in an error as it's non-existent.

For example in the app minimal perms:
https://github.com/directus/directus/blob/aafd63319ef354eeef8c92555c2ba88cdc8fb704/packages/system-data/src/fields/users.yaml#L65-L73

hanneskuettner replied:
Fixed in https://github.com/directus/directus/pull/22773/commits/ad6295310ef67c0999b3046f4e6580bb974f5b65

hanneskuettner committed:
Improve `fetchAllowedFields` to only return fields that are actually in the schema

hanneskuettner committed:
Make a local copy of `junctionFilter` in order to prevent reloads on form value changes

hanneskuettner committed:
Remove debug log

licitdev commented on packages/system-data/src/fields/policies.yaml:
```diff
@@ -0,0 +1,96 @@
+    special:
+      - alias
+      - no-data
+    width: full
```
Hide the `Assigned To` divider for the public role in the policy item form.

```suggestion
    width: full
    <<: *hide_in_public_policy
```

hanneskuettner committed:
Update packages/system-data/src/fields/policies.yaml

Co-authored-by: ian <licitdev@gmail.com>

licitdev commented on api/src/permissions/utils/filter-policies-by-ip.ts:
```diff
@@ -0,0 +1,19 @@
+export function filterPoliciesByIp(policies: AccessRow[], ip: string | null | undefined) {
+	return policies.filter(({ policy }) => {
+		// Keep policies that don't have an ip address allow list configured
+		if (!policy.ip_access) {
```
When all IPs are removed, the `ip_access` field is set to an empty array, resulting in the policy being filtered out.
Note: the typing for `policy.ip_access` is `string[] | null`.

```suggestion
		if (!policy.ip_access || policy.ip_access.length === 0) {
```

hanneskuettner replied:
Good catch!

hanneskuettner committed:
Update api/src/permissions/utils/filter-policies-by-ip.ts

Co-authored-by: ian <licitdev@gmail.com>

hanneskuettner committed:
Conditionally add IP to request level cache key

Add `accountability.ip` to the request level cache key if, and only if, the IP is matched by any of the `ip_access` filters of the current users policies.

This ensures that, if the request IP influences the request result, it is path of the cache key, but also not included if there are no IP filters configured for the current user.

hanneskuettner commented on api/src/permissions/modules/fetch-policies-ip-access/fetch-policies-ip-access.ts:
```diff
@@ -0,0 +1,39 @@
+
+	const rows = await query;
+
+	return rows.map(({ ip_access }) => toArray(ip_access));
```
This need to filter for empty strings.

licitdev replied:
Maybe~~

```suggestion
	return rows
		.map(({ ip_access }) => {
			if (ip_access) return toArray(ip_access);
			return null;
		})
		.filter((val) => val !== null);
```

hanneskuettner replied:
What about?

```suggestion
	return rows
		.filter(({ ip_access }) => ip_access)
		.map(({ ip_access }) => toArray(ip_access));
```

licitdev replied:
Nice! Much neater~ but gotta run the formatted ey 😉 

```suggestion
	return rows.filter(({ ip_access }) => ip_access).map(({ ip_access }) => toArray(ip_access));
```

licitdev replied:
Test comment for @hanneskuettner

hanneskuettner committed:
Update api/src/permissions/modules/fetch-policies-ip-access/fetch-policies-ip-access.ts

Co-authored-by: ian <licitdev@gmail.com>

hanneskuettner committed:
Make sure that fetchAllowedFields does not remove field wildcard '*'

hanneskuettner committed:
Temporarily disable cache key picking

hanneskuettner committed:
Refine cache key picking + remove `undefined` from Accountability['ip'] type

hanneskuettner committed:
Rename `pick` to `prepareArg`

paescuj reviewed (commented):
Not necessarily an issue, but probably something good to be aware of:

If multiple presets exist for the same collection, it will always use the one that was created last, the order of the policies in the roles does not affect this, as one might expect.

<img width="600" src="https://github.com/directus/directus/assets/5363448/21828be1-60d9-434c-a866-3bd2c17cfdea">
<img width="385" src="https://github.com/directus/directus/assets/5363448/c6d40eb5-f5d2-45ee-8aef-1b9bce8614ad">

paescuj commented on api/src/permissions/modules/validate-access/validate-access.ts:
```diff
@@ -0,0 +1,42 @@
+
+	if (!access) {
+		throw new ForbiddenError({
+			reason: `You don't have permission to "${options.action}" from collection "${options.collection}" or it does not exist.`,
```
The message sounds a bit odd
- to "update" from...
- what is "it"?

Maybe something like this makes it a bit clearer:

```suggestion
			reason: `You don't have permission to "${options.action}" on collection "${options.collection}" or the requested resource does not exist.`,
```

hanneskuettner replied:
Keeping it like is for now, because the suggestion doesn't site right with me either '... permission to "read" on collection...' but let's iterate on that in the next one <3

paescuj reviewed (commented):
🚨 Update and delete permissions don't seem to be checked correctly:

<img width="800" src="https://github.com/directus/directus/assets/5363448/6f4ed55a-8369-4317-a72c-4a1accb2adf5">

https://github.com/directus/directus/assets/5363448/fbcc5916-eaee-4ad3-9f04-7d4a324affa9

<img width="800" src="https://github.com/directus/directus/assets/5363448/1c63b3c0-3044-4a2f-9571-d1e76255d1ca">

https://github.com/directus/directus/assets/5363448/c9c454ea-d37c-4126-812c-7b08368e092a

The item is returned in `runAst` of `validate-item-access.ts`, but shouldn't.

hanneskuettner commented:
> 🚨 Update and delete permissions don't seem to be checked correctly:

Are you sure you don't have any policies that accidentally allow full update and delete access to your collection? I just tried replicating this on my instance and for my permissions setup it seems to work as expected:


https://github.com/directus/directus/assets/4376726/dd62ce94-a083-4113-805b-c7b3fc371bf6

hanneskuettner committed:
Define the sort field for the `directus_access` junction table

hanneskuettner commented:
> Not necessarily an issue, but probably something good to be aware of:

> If multiple presets exist for the same collection, it will always use the one that was created last, the order of the policies in the roles does not affect this, as one might expect.

That should be considered a bug, since we want to educate people that the order top to bottom influences the presets. I fixed this by correctly defining the sort field for the collection, as intended.

hanneskuettner committed:
Verify that user has access to automatically selected sort field and default to first allowed field if not

hanneskuettner commented:
> Currently we add "sort":["id"] to the AST in ItemsService.readByQuery which results in not being able to read items for which you dont have read access on id for

@DanielBiegler I've just tackled that in https://github.com/directus/directus/pull/22773/commits/9486a6be24ad2a39b316dbc6dbc5b80378626b2c can you please retest this (is works in my test)

hanneskuettner committed:
Sort the fetched permissions to match the order of the passed in policies

hanneskuettner committed:
Some clean-up of TODOs and unused code

hanneskuettner committed:
Take care of a special case, where no fields are requested and we are still interested the correct items being returned

This surfaced when running `validateItemAccess` with an update permission that did not include the primary key field.

hanneskuettner committed:
3 less

licitdev commented on api/src/database/migrations/20240328A-permissions-policies.ts:
```diff
@@ -0,0 +1,197 @@
+		table.dropColumn('admin_access');
+		table.dropColumn('app_access');
+
+		table.uuid('parent').references('directus_roles.id');
```
There's the foreign key error when a parent role is deleted.
Reminder to rename the migration file to a later date.

```suggestion
		table.uuid('parent').references('directus_roles.id').onDelete('SET NULL');
```

hanneskuettner replied:
> Reminder to rename the migration file to a later date.

Already noted in the top level todos 🫡

paescuj replied:
Let's rename now 😃 I don't expect another migration to sweep in.

hanneskuettner replied:
I didn't do it for now since that would screw up our testing DBs. But I need to touch it later today anyways in order to add the `icon` field. So I might as well do it then.

paescuj replied:
The sooner the better, so we test against the final state 👍

hanneskuettner committed:
Update api/src/database/migrations/20240328A-permissions-policies.ts

Co-authored-by: ian <licitdev@gmail.com>

hanneskuettner committed:
Change /permissions/me response and add corresponding types and constants

paescuj committed:
Fix payload validation

- Field validation needs to happen for admin users as well
- Add back injection of validation rules for non-nullable fields

Tests added/adjusted accordingly

DanielBiegler reviewed (commented):
We seem to not hide the new permissions interface:

![image](https://github.com/directus/directus/assets/14810858/62728db8-c504-4736-991d-8bb3bea4b5d5)

But once I use it, even an admin, simply opening the collection results in FORBIDDEN error:

![image](https://github.com/directus/directus/assets/14810858/15dc45b9-5cb0-47de-9997-37b744498f1d)

Probably best to just hide it from the selectable interfaces?

hanneskuettner commented:
> We seem to not hide the new permissions interface

Huh I didn't even know that was an option. Good catch!

hanneskuettner committed:
Mark `system-permissions` interface as `system`

hanneskuettner committed:
Remove special handling for public policy. It's one of us now.

- Add `icon` field to policies
- Add notice to the public role

hanneskuettner committed:
Rename migration to most recent date

hanneskuettner committed:
Clear permissions cache in `clearSystemCache`

DanielBiegler reviewed (commented):
Indirectly related to: https://github.com/directus/directus/pull/22773#issuecomment-2177799118

Due to sorting via primary key we leak the name of the primary key and existence of collections to users who dont have any read access when doing the following:

| Filter | Fields |
|--------|--------|
| ![image](https://github.com/directus/directus/assets/14810858/1e97dab1-866a-4cae-85d7-d35167b1f5b3) | ![image](https://github.com/directus/directus/assets/14810858/974af9ad-5f6f-4b72-8c7c-fc9f9bfbe420) |  

If an unauthorized user now fetches something we expose the following info:

```
"message": "You don't have permission to access field \"uuid\" in collection \"test\" or it does not exist. Queried in root.",
```

while querying a non-existing table results in:

```
"message":"You don't have permission to access this."
```

Which means unauthorized users can enumerate collections and know the primary key

DanielBiegler reviewed (commented):
The permission interface crashes for me when configuring no rules inside "Use custom"

1. Add any collection to the permission interface be it a user or system one
2. Click on "Use custom"
3. Dont add any rules - no filter, no fields
4. Press save
5. Interface crashes

Video:

[Screencast from 2024-06-19 15-50-55.webm](https://github.com/directus/directus/assets/14810858/7346196c-1296-4787-bb3c-fab3f6c617c8)

paescuj committed:
Ensure fields are always passed to validation-errors

When v-form is used with the `collection` prop, instead of directly
passing fields via `fields` prop, the validation-errors component
didn't receive any field information.
This is fixed, by passing down the "finalFields" from `useForm`.

This is not directly related to 'auditus', but since it seems like this will be the
only place (so far) where we want to show validation-errors on a system
collection, I'm committing here.

hanneskuettner committed:
Make `getDBQuery` not async

hanneskuettner committed:
Set the sort field to `null` if the user does not have any allowed fields, not even the primary key

hanneskuettner committed:
Handle the case where `null` is returned as item edits by the permission detail drawer and remove the existing item, if any.

hanneskuettner committed:
Prevent role recursion (this is a simple check right now and I would expect it to fail on nested role updates that do funky stuff)

hanneskuettner committed:
Add overflow to permissions table

paescuj committed:
Outsource 'useSave' for roles, exactly as in policies

paescuj committed:
Clean-up policies & roles item views

- Remove leftover styles, use clearer naming 'content'
- Fix types (policies was using role type)
- Show validation errors, handle errors on save & delete
- Use loading state of v-form (to have some indicator and less layout shift)

paescuj committed:
Mark name field in policy & role as required

That way, an indicator is shown in the form, and value is checked when
editing via drawer

DanielBiegler committed:
remove role filter from public registration m2o

roles dont have access fields anymore, simply allow admin roles for now

hanneskuettner committed:
Remove overflow again. It broke

hanneskuettner committed:
Add cache purging for permission related updates.

hanneskuettner committed:
Add `dropForeign` in migration

Fixes migration on MySQL

hanneskuettner committed:
Add parsed field name to field map instead of raw field name

This fixes filtering & sorting with function as keys, e.g. `year(date_updated)`

hanneskuettner committed:
Account for $FOLLOW field filters earlier and don't confuse them with functions (see prev commit)

hanneskuettner committed:
Update migration to also work with CockroachDB

Instead of altering the `policy` column on `directus_permissions` to add the NOT NULL constraint it needs to be created with NOT NULL in the first place, as this will fail in CockroachDB.

That means we need to drop the foreign key constraint for the role column in order to update to `null` `role` to the public policy ID before we copy the values into the `policy` column

hanneskuettner committed:
Fix typo

hanneskuettner committed:
Add icon to default admin policy

licitdev commented on api/src/permissions/lib/fetch-policies.ts:
```diff
@@ -0,0 +1,39 @@
+
+	let filter: Filter;
+
+	// No roles and no user means unauthenticated request
+	if (roles.length === 0 && !user) {
```
Users without a role no longer assume the Public role. Should this behaviour be changed?

```suggestion
	// Public role is used when the role or user is missing
	if (roles.length === 0 || !user) {
```

licitdev replied:
Updated suggestion to include policies directly attached to users.

```suggestion
	if (roles.length === 0) {
		// Users without role assumes the Public role permissions along with their attached policies
		if (user) {
			filter = { _or: [{ role: { _null: true }, user: { _null: true } }, { user: { _eq: user } }] };
		}
		// No roles and no user means an unauthenticated request
		else {
			filter = { role: { _null: true }, user: { _null: true } };
		}
```

hanneskuettner committed:
Be more clear about where the public role applies

hanneskuettner committed:
Update api/src/permissions/lib/fetch-policies.ts

Co-authored-by: ian <licitdev@gmail.com>

licitdev commented on api/src/permissions/lib/fetch-policies.test.ts:
```diff
@@ -0,0 +1,107 @@
+
+	AccessService.prototype.readByQuery = vi.fn().mockResolvedValue(rows);
+});
+
```
Update test for changes in https://github.com/directus/directus/pull/22773/files#r1647286436

```suggestion

test('Fetches policies for public role and user when user is given without role', async () => {
	const acc = { roles: [], user: 'user-a' } as unknown as Accountability;

	const policies = await fetchPolicies(acc, {} as Context);

	expect(AccessService.prototype.readByQuery).toHaveBeenCalledWith({
		filter: {
			_or: [
				{
					role: {
						_null: true,
					},
					user: {
						_null: true,
					},
				},
				{ user: { _eq: 'user-a' } },
			],
		},
		fields: ['policy.id', 'policy.ip_access'],
		limit: -1,
	});

	expect(policies).toEqual([]);
});

```

hanneskuettner committed:
Update api/src/permissions/lib/fetch-policies.test.ts

Co-authored-by: ian <licitdev@gmail.com>

licitdev committed:
Enable GH workflows

hanneskuettner committed:
Make eslint happy, cleanup and improve role sidebar in users view

hanneskuettner committed:
Fix wrong suggestion application and move comments in block

paescuj committed:
Format files

paescuj committed:
Update isFullPermission test

hanneskuettner committed:
Clean up policy filter logic

hanneskuettner committed:
Flip order in test

licitdev commented on packages/system-data/src/fields/policies.yaml:
```diff
@@ -0,0 +1,89 @@
+    interface: list-m2m
+    options:
+      template: '{{ role.name }}'
+      junctionFilter:
+        role:
+          _nnull: true
```
The M2M interface does not show the public role when it's attached.
<img width="863" alt="image" src="https://github.com/directus/directus/assets/26413686/8cb68ca1-d47a-4923-9a94-ecd587065c08">

If the junction filter is removed, the interface will show an empty item for the public role. 🤔 

```suggestion
```

<img width="824" alt="image" src="https://github.com/directus/directus/assets/26413686/f1d4844a-b6ef-4c42-a1c6-18b6a6caccc2">

<img width="834" alt="image" src="https://github.com/directus/directus/assets/26413686/c68a9814-07f2-4c79-b494-1fe5288e4775">

paescuj replied:
Hmm is the public role missing some props?

hanneskuettner replied:
The public role does not exist at all, right? It is a virtual role that is composed of all policies that have `role: null` and `user: null` in `directus_access`

hanneskuettner replied:
So that is to be expected for now. Maybe we can convince the M2M to show this virtual row as well, but not really in the scope of the RC I'd say.

paescuj committed:
Update mocking in user/flows store tests

hanneskuettner committed:
Update expected results of a lot of api tests that changed during development (all stay true to the original idea)

paescuj committed:
Update parseFilter test

hanneskuettner committed:
Update injectCases test

DanielBiegler reviewed (commented):
Found some weird behavior which denies you from changing any users roles after you look them up through a relation once

1. Log into your admin account
2. Look up your user, your role field is editable as it should be
3. Now open Settings > User Roles > Admin Role
4. At the bottom in the "Users in role" click on your admin user so that the drawer opens
5. Scroll down to the roles field to confirm its un-editable
6. Close the drawer without doing any changes
7. Now you cant change any roles for any user in the UI because the roles-field is disabled on each users item detail page be it admin or non-admin

Breaks the same way if you look up a non-admin user from the "User Roles" view. The video shows this.

Here a video:

[Screencast from 2024-06-20 14-13-16.webm](https://github.com/directus/directus/assets/14810858/3ae305d4-0719-481a-9464-5c2855bc7066)

hanneskuettner committed:
Manually set parent to `null` if a role gets deleted and remove the `SET NULL` on delete action.

The `SET NULL` action causes problems problems on OracleDB

paescuj commented:
> Found some weird behavior which denies you from changing any users roles after you look them up through a relation once
> 
>     1. Log into your admin account
> 
>     2. Look up your user, your role field is editable as it should be
> 
>     3. Now open Settings > User Roles > Admin Role
> 
>     4. At the bottom in the "Users in role" click on your admin user so that the drawer opens
> 
>     5. Scroll down to the roles field to confirm its un-editable
> 
>     6. Close the drawer without doing any changes
> 
>     7. Now you cant change any roles for any user in the UI because the roles-field is disabled on each users item detail page be it admin or non-admin
> 
> 
> Breaks the same way if you look up a non-admin user from the "User Roles" view. The video shows this.
> 
> Here a video:
> Screencast.from.2024-06-20.14-13-16.webm

This is actually unrelated to the changes in this pull request - has been an issue before 😢

hanneskuettner commented:
> Found some weird behavior which denies you from changing any users roles after you look them up through a relation once

Fun one. That is also a reproducible problem in the current `main`. Probably caused by the details drawer locking down the junction field and doing that on the fields returned by the field store instead of on a local copy. I'll open an issue for it so it can be fixed.

paescuj committed:
Fix limit check for new users w/o "status" field

status defaults to "active", thus if the field is not in payload, the
user limit check needs to be triggered

br41nslug committed:
Fixed migration "inconsistent datatypes: expected - got CLOB" error in oracle

br41nslug committed:
Merge branch 'auditus' of https://github.com/directus/directus into auditus

hanneskuettner committed:
Update extractFieldsFromChildren test

hanneskuettner committed:
Fix `count(o2m)` type queries

paescuj committed:
Update UsersService tests

- Adapt to new user integrity logic
- Add basic ItemsService tests to ensure user integrity checks take
  place

paescuj committed:
Move withAppMinimalPermissions to appropriate dir

hanneskuettner committed:
Make sure that relational function aliases are recognized as `functionField`

br41nslug committed:
Fixed boolean logic error for graphql counting

hanneskuettner committed:
Fix permission for relational functions

Before this functions that operated on a o2m field like `count(o2m)` did not respect permissions on the related collection.

Now function field nodes have a cases list as well and correctly get the cases injected for the related collection.
The permissions are then correctly injected in the query that is passed down to the relational count function helper.

hanneskuettner committed:
Fix mock in withAppMinimalPermissions test

hanneskuettner committed:
Merge branch 'v11-rc' into auditus

hanneskuettner committed:
Add preliminary changesets

paescuj committed:
Update RolesService tests

Copied and commented out old checks from roles to policies, so we can
re-check later on

hanneskuettner committed:
Reword changeset to "Policies"

paescuj commented on .changeset/strong-numbers-warn.md:
```diff
@@ -0,0 +1,5 @@
+---
+'@directus/errors': major
```
Not a breaking change, IMO
```suggestion
'@directus/errors': minor
```

hanneskuettner replied:
Oh yes. I was under the impression that `reason` is a required argument now. Minor it is.

hanneskuettner committed:
Update .changeset/strong-numbers-warn.md

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

phazonoverload committed:
Policies Documentation (#22729)

Co-authored-by: Hannes Küttner <kuettner.hannes@gmail.com>
Co-authored-by: Brainslug <tim@brainslug.nl>

hanneskuettner committed:
Reformat docs

br41nslug committed:
SDK functions for auditus (#22795)

* Updated sdk types

* added policy commands

* prettier

* Added new and missing websocket subscription hooks

* Added missing endpoints

* Added missing graphql endpoints

* prettier

* Added changesets

* updated changesets

* Update .changeset/nine-geckos-jog.md

* Update api/src/services/graphql/index.ts

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

* Update sdk/src/rest/commands/create/policies.ts

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

* Update sdk/src/rest/commands/create/policies.ts

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

* Update sdk/src/rest/commands/read/policies.ts

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

* Update sdk/src/rest/commands/read/roles.ts

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

* Update sdk/src/rest/commands/read/policies.ts

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

* Update sdk/src/rest/commands/read/policies.ts

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

* Update sdk/src/rest/commands/read/policies.ts

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

* Update sdk/src/schema/policy.ts

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

* Update sdk/src/schema/policy.ts

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

---------

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

br41nslug committed:
fixed icon type

hanneskuettner committed:
Added app & api changeset
